File: examples/onedheat/onedheat.py
#!/usr/bin/env python
from __future__ import print_function
from numba import jit
import numpy as np
import argparse
import time

@jit('f8[:](i4, i4, f8, f8, f8, f8, f8)', nopython=True)
def onedheat(ngrid, ntimesteps, kappa, xleft, xright, tleft, tright):
    dx = (xright-xleft)/(ngrid-1)
    dt = 0.25*dx*dx/kappa

    temp = np.zeros(ngrid+2, dtype=np.double)
    temp_new = np.zeros(ngrid+2, dtype=np.double)
    temp[0], temp[ngrid+1] = tleft, tright

    for iteration in range(ntimesteps):
        temp_new[1:ngrid] = temp[1:ngrid] + kappa*dt/(dx*dx)*\
            (temp[2:ngrid+1] - 2.*temp[1:ngrid] + temp[0:ngrid-1])

        temp[1:ngrid] = temp_new[1:ngrid]

    return temp[1:ngrid]

if __name__ == "__main__":
    parser = argparse.ArgumentParser()
    parser.add_argument('-n', '--ntimesteps', type=np.int, default=10000)
    parser.add_argument('-g', '--ngrid', type=np.int, default=1001)
    parser.add_argument('-k', '--kappa', type=np.double, default=1.0)
    parser.add_argument('-L', '--xleft', type=np.double, default=0.)
    parser.add_argument('-R', '--xright', type=np.double, default=1.)
    parser.add_argument('-l', '--tleft', type=np.double, default=-1.)
    parser.add_argument('-r', '--tright', type=np.double, default=+1.)
    args = parser.parse_args()

    start = time.clock()
    temps = onedheat(args.ngrid, args.ntimesteps, args.kappa, args.xleft, args.xright, args.tleft, args.tright)
    print(time.clock() - start)
    print(temps[args.ngrid//2])

File: onedheat/onedheat.jl
function main(ngrid, ntimesteps, kappa=1.0, xleft=0.0, xright=1.0, tleft=-1.0, tright=1.0)
  const dx = (xright-xleft)/(ngrid-1)
  const dt = 0.25dx * dx / kappa

  temp = Array{Float64}(ngrid+2) 
  temp_new = Array{Float64}(ngrid+2) 
  
  for i in 2:ngrid+1
    @inbounds temp[i] = 0.
  end

  temp[1] = tleft
  temp[ngrid+2] = tright

  for iteration in 1:ntimesteps
    for i in 2:ngrid+1
        @inbounds temp_new[i] = temp[i] + kappa*dt/(dx*dx)*(temp[i-1] - 2*temp[i] + temp[i+1])
    end
    for i in 2:ngrid+1
        @inbounds temp[i] = temp_new[i]
    end
  end

  println(temp[div(ngrid,2)+2])
end

ngrid = 1001
ntimesteps = 10000
nargs = length(ARGS)
if nargs > 0
    ngrid = parse(Int32, ARGS[1])
end
if nargs > 1
    ntimesteps = parse(Int32, ARGS[2])
end
main(10, 10)
@time main(ngrid, ntimesteps)
